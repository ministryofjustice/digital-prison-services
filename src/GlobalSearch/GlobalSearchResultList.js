import React from 'react'
import '../index.scss'
import '../lists.scss'
import '../App.scss'
import PropTypes from 'prop-types'
import { getOffenderLink } from '../links'
import PreviousNextNavigation from '../PreviousNextNavigation'
import { properCaseName } from '../utils'
import ResultsFilter from '../Components/ResultsFilter'

const GlobalSearchResultList = ({
  data,
  pageSize,
  pageNumber,
  totalRecords,
  handlePageAction,
  licencesUser,
  licencesVaryUser,
  licencesUrl,
  searchPerformed,
  viewInactivePrisoner,
  handleResultsPerPageChange,
}) => {
  const headings = (
    <tr>
      {/* eslint-disable-next-line jsx-a11y/control-has-associated-label */}
      <th className="straight" />
      <th className="straight">Name</th>
      <th className="straight">Prison&nbsp;no.</th>
      <th className="straight">Date of birth</th>
      <th className="straight">Location</th>
      <th className="straight">Actual working name</th>
      {/* eslint-disable-next-line jsx-a11y/control-has-associated-label */}
      {licencesUser && <th className="straight" />}
    </tr>
  )

  const offenderImageUrl = offenderNo => `/app/images/${offenderNo}/data`

  const canUpdateLicence = prisoner => licencesUser && (prisoner.currentlyInPrison === 'Y' || licencesVaryUser)

  const offenders =
    data &&
    data.map(prisoner => {
      const shouldAddLink =
        (viewInactivePrisoner && prisoner.currentlyInPrison === 'N') || prisoner.currentlyInPrison === 'Y'
      return (
        <tr key={`${prisoner.offenderNo}-${prisoner.uiId}`} className="row-gutters">
          <td className="row-gutters">
            {shouldAddLink ? (
              <a
                id={`imageLink-${prisoner.offenderNo}`}
                target="_blank"
                rel="noopener noreferrer"
                className="link"
                href={getOffenderLink(prisoner.offenderNo)}
              >
                <img
                  id={`image-${prisoner.offenderNo}`}
                  alt={`prisoner ${prisoner.offenderNo}`}
                  className="clickable"
                  height="100"
                  width="80"
                  src={offenderImageUrl(prisoner.offenderNo)}
                />
              </a>
            ) : (
              <img
                alt={`prisoner ${prisoner.offenderNo}`}
                id={`image-${prisoner.offenderNo}`}
                height="100"
                width="80"
                src={offenderImageUrl(prisoner.offenderNo)}
              />
            )}
          </td>
          {shouldAddLink ? (
            <td className="row-gutters">
              <a target="_blank" rel="noopener noreferrer" className="link" href={getOffenderLink(prisoner.offenderNo)}>
                {properCaseName(prisoner.lastName)}, {properCaseName(prisoner.firstName)}
              </a>
            </td>
          ) : (
            <td className="row-gutters">
              {properCaseName(prisoner.lastName)}, {properCaseName(prisoner.firstName)}
            </td>
          )}
          <td className="row-gutters">{prisoner.offenderNo}</td>
          <td className="row-gutters">{prisoner.dateOfBirth}</td>
          <td className="row-gutters">{prisoner.latestLocation}</td>
          <td className="row-gutters">
            {properCaseName(prisoner.currentWorkingLastName)}, {properCaseName(prisoner.currentWorkingFirstName)}
          </td>
          {canUpdateLicence(prisoner) && (
            <td className="row-gutters">
              <a
                href={`${licencesUrl}hdc/taskList/${prisoner.latestBookingId}`}
                className="clear-filters link clickable toLicences"
              >
                Update licence
              </a>
            </td>
          )}
        </tr>
      )
    })

  const pagination = { perPage: pageSize, pageNumber }
  const noResultsText = searchPerformed ? 'No prisoners found' : 'Use the search box above'

  return (
    <div>
      {totalRecords > 0 && (
        <ResultsFilter perPage={pageSize} pageNumber={pageNumber} totalResults={totalRecords} noBorder>
          <ResultsFilter.PerPageDropdown
            handleChange={handleResultsPerPageChange}
            totalResults={totalRecords}
            perPage={pageSize}
          />
        </ResultsFilter>
      )}
      <table className="row-gutters">
        <thead>{headings}</thead>
        <tbody>{offenders}</tbody>
      </table>
      {(!offenders || offenders.length === 0) && (
        <div className="font-small padding-top-large padding-bottom padding-left">{noResultsText}</div>
      )}
      <div>
        <PreviousNextNavigation
          pagination={pagination}
          totalRecords={totalRecords}
          pageAction={id => {
            handlePageAction(id)
          }}
        />
      </div>
    </div>
  )
}

GlobalSearchResultList.propTypes = {
  // props
  handlePageAction: PropTypes.func.isRequired,
  data: PropTypes.arrayOf(
    PropTypes.shape({
      offenderNo: PropTypes.string.isRequired,
      lastName: PropTypes.string.isRequired,
      firstName: PropTypes.string.isRequired,
      dateOfBirth: PropTypes.string.isRequired,
      latestLocation: PropTypes.string.isRequired,
      uiId: PropTypes.string,
    })
  ).isRequired,
  pageNumber: PropTypes.number.isRequired,
  pageSize: PropTypes.number.isRequired,
  totalRecords: PropTypes.number.isRequired,
  licencesUser: PropTypes.bool.isRequired,
  licencesVaryUser: PropTypes.bool.isRequired,
  viewInactivePrisoner: PropTypes.bool.isRequired,
  licencesUrl: PropTypes.string.isRequired,
  searchPerformed: PropTypes.bool.isRequired,
  handleResultsPerPageChange: PropTypes.func.isRequired,
}

export default GlobalSearchResultList
