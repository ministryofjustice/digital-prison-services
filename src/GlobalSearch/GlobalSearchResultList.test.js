import React from 'react'
import { shallow } from 'enzyme'
import GlobalSearchResultList from './GlobalSearchResultList'

const OFFENDER_NAME_COLUMN = 0
const NOMS_ID_COLUMN = 1
const DOB_COLUMN = 2
const LOCATION_COLUMN = 3

const response = [
  {
    offenderNo: 'A1234AC',
    firstName: 'FRED',
    lastName: 'QUIMBY',
    latestLocation: 'LEEDS HMP',
    dateOfBirth: '1977-10-15',
  },
  {
    offenderNo: 'A1234AA',
    firstName: 'ARTHUR',
    lastName: 'ANDERSON',
    latestLocation: 'Moorland HMP',
    dateOfBirth: '1976-09-15',
  },
]

describe('Global search results component', () => {
  it('should render results correctly', async () => {
    const component = shallow(<GlobalSearchResultList data={response} />)

    const tr = component.find('tr')
    expect(tr.length).toEqual(3) // 2 plus table header tr
    expect(
      tr
        .at(1)
        .find('td a')
        .at(OFFENDER_NAME_COLUMN)
        .text()
    ).toEqual('Quimby, Fred')
    expect(
      tr
        .at(1)
        .find('td')
        .at(NOMS_ID_COLUMN)
        .text()
    ).toEqual('A1234AC')
    expect(
      tr
        .at(1)
        .find('td')
        .at(LOCATION_COLUMN)
        .text()
    ).toEqual('LEEDS HMP')
    expect(
      tr
        .at(1)
        .find('td')
        .at(DOB_COLUMN)
        .text()
    ).toEqual('1977-10-15')

    expect(
      tr
        .at(2)
        .find('td a')
        .at(OFFENDER_NAME_COLUMN)
        .text()
    ).toEqual('Anderson, Arthur')
    expect(
      tr
        .at(2)
        .find('td')
        .at(NOMS_ID_COLUMN)
        .text()
    ).toEqual('A1234AA')
    expect(
      tr
        .at(2)
        .find('td')
        .at(LOCATION_COLUMN)
        .text()
    ).toEqual('Moorland HMP')
    expect(
      tr
        .at(2)
        .find('td')
        .at(DOB_COLUMN)
        .text()
    ).toEqual('1976-09-15')
  })

  it('should render empty results list correctly', async () => {
    const component = shallow(<GlobalSearchResultList data={[]} />)
    const tr = component.find('tr')
    expect(tr.length).toEqual(1) // table header tr only
    expect(component.find('div.font-small').text()).toEqual('No prisoners found')
  })
})
