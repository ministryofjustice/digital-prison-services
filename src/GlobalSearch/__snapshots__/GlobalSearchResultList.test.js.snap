// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Global search results component should render results correctly 1`] = `
<div>
  <div
    className="sc-htoDjs fjTtnT"
  >
    <div
      className="sc-gzVnrw jpnZQI"
    >
      4
       - 
      4
       of
       
      4
       results
    </div>
    <div
      className="sc-jTzLTM dwYLEN"
    >
      <label
        htmlFor="perPage"
      >
        <span
          className="sc-dnqmqq jvzDFl"
        >
          Results per page
        </span>
      </label>
      <select
        className="sc-VigVT fNFOws"
        id="perPage"
        onChange={[Function]}
        value={4}
      >
        <option
          value={4}
        >
          All
        </option>
      </select>
    </div>
  </div>
  <table
    className="row-gutters"
  >
    <thead>
      <tr>
        <th
          className="straight"
        />
        <th
          className="straight"
        >
          Name
        </th>
        <th
          className="straight"
        >
          Prison no.
        </th>
        <th
          className="straight"
        >
          Date of birth
        </th>
        <th
          className="straight"
        >
          Location
        </th>
        <th
          className="straight"
        >
          Actual working name
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        className="row-gutters"
      >
        <td
          className="row-gutters"
        >
          <a
            className="link"
            href="/prisoner/A1234AC"
            id="imageLink-A1234AC"
            rel="noopener noreferrer"
            target="_blank"
          >
            <img
              alt="prisoner A1234AC"
              className="clickable"
              height="100"
              id="image-A1234AC"
              src="/app/images/A1234AC/data"
              width="80"
            />
          </a>
        </td>
        <td
          className="row-gutters"
        >
          <a
            className="link"
            href="/prisoner/A1234AC"
            rel="noopener noreferrer"
            target="_blank"
          >
            Quimby
            , 
            Fred
          </a>
        </td>
        <td
          className="row-gutters"
        >
          A1234AC
        </td>
        <td
          className="row-gutters"
        >
          1977-10-15
        </td>
        <td
          className="row-gutters"
        >
          LEEDS HMP
        </td>
        <td
          className="row-gutters"
        >
          
          , 
          
        </td>
      </tr>
      <tr
        className="row-gutters"
      >
        <td
          className="row-gutters"
        >
          <a
            className="link"
            href="/prisoner/A1234AA"
            id="imageLink-A1234AA"
            rel="noopener noreferrer"
            target="_blank"
          >
            <img
              alt="prisoner A1234AA"
              className="clickable"
              height="100"
              id="image-A1234AA"
              src="/app/images/A1234AA/data"
              width="80"
            />
          </a>
        </td>
        <td
          className="row-gutters"
        >
          <a
            className="link"
            href="/prisoner/A1234AA"
            rel="noopener noreferrer"
            target="_blank"
          >
            Anderson
            , 
            Arthur
          </a>
        </td>
        <td
          className="row-gutters"
        >
          A1234AA
        </td>
        <td
          className="row-gutters"
        >
          1976-09-15
        </td>
        <td
          className="row-gutters"
        >
          Moorland HMP
        </td>
        <td
          className="row-gutters"
        >
          
          , 
          
        </td>
      </tr>
      <tr
        className="row-gutters"
      >
        <td
          className="row-gutters"
        >
          <img
            alt="prisoner A1234AB"
            height="100"
            id="image-A1234AB"
            src="/app/images/A1234AB/data"
            width="80"
          />
        </td>
        <td
          className="row-gutters"
        >
          Runaway
          , 
          Peter
        </td>
        <td
          className="row-gutters"
        >
          A1234AB
        </td>
        <td
          className="row-gutters"
        >
          1970-09-15
        </td>
        <td
          className="row-gutters"
        >
          Released From Leeds
        </td>
        <td
          className="row-gutters"
        >
          
          , 
          
        </td>
      </tr>
    </tbody>
  </table>
  <div>
    <nav
      className="sc-bdVaJa fGYvF"
    >
      <ul
        className="sc-bwzfXH iHMVMS"
      >
        <li
          className="sc-htpNat bYwZFO"
          id="previous-page"
          onClick={[Function]}
          role="button"
        >
          <div
            className="sc-bxivhb lhbcqk"
          >
            Previous
            <span
              className="sc-EHOje dcPHNN"
            >
              page
            </span>
            <span
              className="sc-bZQynM kFuPMl"
            >
              1
               of 
              1
            </span>
          </div>
        </li>
        <li
          className="sc-ifAKCX jEPezK"
        />
      </ul>
    </nav>
  </div>
</div>
`;

exports[`Global search results component should render results when has inactive booking role 1`] = `
<div>
  <div
    className="sc-htoDjs fjTtnT"
  >
    <div
      className="sc-gzVnrw jpnZQI"
    >
      4
       - 
      4
       of
       
      4
       results
    </div>
    <div
      className="sc-jTzLTM dwYLEN"
    >
      <label
        htmlFor="perPage"
      >
        <span
          className="sc-dnqmqq jvzDFl"
        >
          Results per page
        </span>
      </label>
      <select
        className="sc-VigVT fNFOws"
        id="perPage"
        onChange={[Function]}
        value={4}
      >
        <option
          value={4}
        >
          All
        </option>
      </select>
    </div>
  </div>
  <table
    className="row-gutters"
  >
    <thead>
      <tr>
        <th
          className="straight"
        />
        <th
          className="straight"
        >
          Name
        </th>
        <th
          className="straight"
        >
          Prison no.
        </th>
        <th
          className="straight"
        >
          Date of birth
        </th>
        <th
          className="straight"
        >
          Location
        </th>
        <th
          className="straight"
        >
          Actual working name
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        className="row-gutters"
      >
        <td
          className="row-gutters"
        >
          <a
            className="link"
            href="/prisoner/A1234AC"
            id="imageLink-A1234AC"
            rel="noopener noreferrer"
            target="_blank"
          >
            <img
              alt="prisoner A1234AC"
              className="clickable"
              height="100"
              id="image-A1234AC"
              src="/app/images/A1234AC/data"
              width="80"
            />
          </a>
        </td>
        <td
          className="row-gutters"
        >
          <a
            className="link"
            href="/prisoner/A1234AC"
            rel="noopener noreferrer"
            target="_blank"
          >
            Quimby
            , 
            Fred
          </a>
        </td>
        <td
          className="row-gutters"
        >
          A1234AC
        </td>
        <td
          className="row-gutters"
        >
          1977-10-15
        </td>
        <td
          className="row-gutters"
        >
          LEEDS HMP
        </td>
        <td
          className="row-gutters"
        >
          
          , 
          
        </td>
      </tr>
      <tr
        className="row-gutters"
      >
        <td
          className="row-gutters"
        >
          <a
            className="link"
            href="/prisoner/A1234AA"
            id="imageLink-A1234AA"
            rel="noopener noreferrer"
            target="_blank"
          >
            <img
              alt="prisoner A1234AA"
              className="clickable"
              height="100"
              id="image-A1234AA"
              src="/app/images/A1234AA/data"
              width="80"
            />
          </a>
        </td>
        <td
          className="row-gutters"
        >
          <a
            className="link"
            href="/prisoner/A1234AA"
            rel="noopener noreferrer"
            target="_blank"
          >
            Anderson
            , 
            Arthur
          </a>
        </td>
        <td
          className="row-gutters"
        >
          A1234AA
        </td>
        <td
          className="row-gutters"
        >
          1976-09-15
        </td>
        <td
          className="row-gutters"
        >
          Moorland HMP
        </td>
        <td
          className="row-gutters"
        >
          
          , 
          
        </td>
      </tr>
      <tr
        className="row-gutters"
      >
        <td
          className="row-gutters"
        >
          <a
            className="link"
            href="/prisoner/A1234AB"
            id="imageLink-A1234AB"
            rel="noopener noreferrer"
            target="_blank"
          >
            <img
              alt="prisoner A1234AB"
              className="clickable"
              height="100"
              id="image-A1234AB"
              src="/app/images/A1234AB/data"
              width="80"
            />
          </a>
        </td>
        <td
          className="row-gutters"
        >
          <a
            className="link"
            href="/prisoner/A1234AB"
            rel="noopener noreferrer"
            target="_blank"
          >
            Runaway
            , 
            Peter
          </a>
        </td>
        <td
          className="row-gutters"
        >
          A1234AB
        </td>
        <td
          className="row-gutters"
        >
          1970-09-15
        </td>
        <td
          className="row-gutters"
        >
          Released From Leeds
        </td>
        <td
          className="row-gutters"
        >
          
          , 
          
        </td>
      </tr>
    </tbody>
  </table>
  <div>
    <nav
      className="sc-bdVaJa fGYvF"
    >
      <ul
        className="sc-bwzfXH iHMVMS"
      >
        <li
          className="sc-htpNat bYwZFO"
          id="previous-page"
          onClick={[Function]}
          role="button"
        >
          <div
            className="sc-bxivhb lhbcqk"
          >
            Previous
            <span
              className="sc-EHOje dcPHNN"
            >
              page
            </span>
            <span
              className="sc-bZQynM kFuPMl"
            >
              1
               of 
              1
            </span>
          </div>
        </li>
        <li
          className="sc-ifAKCX jEPezK"
        />
      </ul>
    </nav>
  </div>
</div>
`;
